lib/Parse/ParseType.cpp

```diff
@@ -182,59 +183,46 @@ ParserResult<TypeRepr> Parser::parseType(Diag<> MessageID,
   if (ty.isNull())
     return nullptr;
 
-  // Parse a throws specifier.  'throw' is probably a typo for 'throws',
-  // but in local contexts we could just be at the end of a statement,
-  // so we need to check for the arrow.
-  ParserPosition beforeThrowsPos;
+  // Parse a throws specifier.
+  // Don't consume 'throws', if next token is not '->', so we can emit a more
+  // useful diagnostic when parsing a function decl.
   SourceLoc throwsLoc;
-  bool rethrows = false;
-  if (Tok.isAny(tok::kw_throws, tok::kw_rethrows) ||
-      (Tok.is(tok::kw_throw) && peekToken().is(tok::arrow))) {
-    if (Tok.is(tok::kw_throw)) {
+  if (Tok.isAny(tok::kw_throws, tok::kw_rethrows, tok::kw_throw) &&
+      peekToken().is(tok::arrow)) {
```

Comment:
I'm not sure. What you mention is `Don't consume 'throws', if next token is not '->'`?
Swift 3.0.1 and this PR both emits:
```
test.swift:1:11: error: consecutive statements on a line must be separated by ';'
let fn: () throws
          ^
          ;
test.swift:1:12: error: expected expression
let fn: () throws
           ^
```
Actually, this behavior is from the original code:
https://github.com/apple/swift/pull/5721/files#diff-3d57a567ce2b696aa81ceb836f678fdeL222